File: backend/app/models/employee.py
from sqlalchemy import Date, DateTime, Index, Integer, Numeric, String
from sqlalchemy.sql import func
from sqlalchemy.orm import mapped_column, Mapped

from app.database import Base


class Employee(Base):
    __tablename__ = "employees"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    employee_id: Mapped[str] = mapped_column(String(20), unique=True, nullable=False)
    full_name: Mapped[str] = mapped_column(String(255), nullable=False)
    email: Mapped[str] = mapped_column(String(255), unique=True, nullable=False)
    job_title: Mapped[str] = mapped_column(String(100), nullable=False)
    department: Mapped[str] = mapped_column(String(100), nullable=False)
    country: Mapped[str] = mapped_column(String(100), nullable=False)
    salary: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    currency: Mapped[str] = mapped_column(String(10), nullable=False, default="USD")
    employment_type: Mapped[str] = mapped_column(String(50), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="Active")
    hire_date: Mapped[Date] = mapped_column(Date, nullable=False)
    created_at: Mapped[DateTime] = mapped_column(DateTime, nullable=False, default=func.now())
    updated_at: Mapped[DateTime] = mapped_column(
        DateTime,
        nullable=False,
        default=func.now(),
        onupdate=func.now(),
    )


Index("ix_employees_country", Employee.country)
Index("ix_employees_job_title", Employee.job_title)
Index("ix_employees_department", Employee.department)
Index("ix_employees_status", Employee.status)
Index("ix_employees_salary", Employee.salary)

File: frontend/src/types/employee.ts
export interface Employee {
  id: number;
  employee_id: string;
  full_name: string;
  email: string;
  job_title: string;
  department: string;
  country: string;
  salary: number;
  currency: string;
  employment_type: string;
  status: string;
  hire_date: string;
  created_at: string;
  updated_at: string;
}

export interface EmployeeCreate {
  full_name: string;
  email: string;
  job_title: string;
  department: string;
  country: string;
  salary: number;
  currency: string;
  employment_type: string;
  status: string;
  hire_date: string;
}

export type EmployeeUpdate = EmployeeCreate;
export type EmployeePatch = Partial<EmployeeCreate>;

export interface Pagination {
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface PaginatedResponse<T> {
  data: T[];
  pagination: Pagination;
}

export interface EmployeeListParams {
  page?: number;
  page_size?: number;
  search?: string;
  country?: string;
  department?: string;
  job_title?: string;
  status?: string;
  employment_type?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}
